skyve-war/src/main/java/modules/admin/UserLoginRecord/UserLoginRecordBizlet.java

```diff
@@ -0,0 +1,129 @@
+
+	public static final String COMMUNICATION_DESCRIPTION = "Email warning a user of a new login from a different country";
+
+	/**
+	 * The preSave is overridden so as to add the country of the user based on the ip addrress and if IpInfo token is set. The
+	 * method also adds security logs if the ip address and/or the country code of the user changed from the previous login. It also
+	 * kicks off a job sending the user an email if the country changed from the previous login.
+	 */
+	@Override
+	public void preSave(UserLoginRecordExtension bean) throws Exception {
+
+		String country = bean.getCountry();
+		// Check if the IpInfo token has been set so as to get the country code and country
+		if (UtilImpl.IP_INFO_TOKEN != null) {
+
+			Optional<String> countryCodeOptional = geoIPService.getCountryCodeForIP(bean.getIpAddress());
+			if (countryCodeOptional.isPresent()) {
+				String countryCode = countryCodeOptional.get();
+				Locale locale = new Locale("", countryCode);
+				country = locale.getDisplayCountry();
+				bean.setCountry(country);
+			}
+		}
+
+		// Get the previous login record of the current user
+		DocumentQuery q = CORE.getPersistence()
+				.newDocumentQuery(UserLoginRecord.MODULE_NAME, UserLoginRecord.DOCUMENT_NAME);
+		q.getFilter()
+				.addEquals(AppConstants.USER_NAME_ATTRIBUTE_NAME, bean.getUserName());
+		q.addBoundOrdering(UserLoginRecord.loginDateTimePropertyName, SortDirection.descending);
+		UserLoginRecordExtension previousLoginRecord = q.beanResult();
+
+		// If there is no previous record then this is probably their first login to the system
+		if (previousLoginRecord != null) {
+			String userIPAddress = bean.getIpAddress();
+
+			// Get admin user by filtering based on the username due to usernames being unique
+			DocumentQuery qUser = CORE.getPersistence()
+					.newDocumentQuery(User.MODULE_NAME, User.DOCUMENT_NAME);
+			qUser.getFilter()
+					.addEquals(User.userNamePropertyName, bean.getUserName());
+			UserExtension adminUser = qUser.beanResult();
+
+			if (adminUser != null) {
+				final Persistence persistence = CORE.getPersistence();
+				final org.skyve.metadata.user.User user = persistence.getUser();
+				final Customer customer = user.getCustomer();
+				final Module module = customer.getModule(User.MODULE_NAME);
+				final JobMetaData countryChangeNotificationJobMetaData = module
+						.getJob(DifferentCountryLoginNotificationJob.JOB_NAME);
+				// Get username
+				String userName = adminUser.getContact()
+						.getName();
+
+				// Check if the ip address changed since last login and if so log the event
+				String lastIpAddress = (String) BindUtil.get(previousLoginRecord, IP_ADDRESS);
+				if (lastIpAddress != null && !Objects.equals(userIPAddress, lastIpAddress)) {
+
+					// Check if the country has changed since the last login and if so send the user a warning message
+					String previousCountry = (String) BindUtil.get(previousLoginRecord, COUNTRY);
```

Comment:
You can user a getter here as you are no longer restricted to an `AbstractPersistentBean`.